fem/libceed/algebraic.cpp

```diff
@@ -0,0 +1,620 @@
+   {
+      mem = CEED_MEM_HOST;
+   }
+   Vector local_diag(length);
+   CeedScalar *ptr = (mem == CEED_MEM_HOST) ? local_diag.HostWrite() :
+                     local_diag.Write(true);
+   CeedVectorSetArray(diagceed, mem, CEED_USE_POINTER, ptr);
+   CeedOperatorLinearAssembleDiagonal(ceed_op, diagceed, CEED_REQUEST_IMMEDIATE);
+   CeedVectorTakeArray(diagceed, mem, NULL);
+
+   Vector t_diag;
+   if (P)
+   {
+      t_diag.SetSize(P->Width());
+      P->MultTranspose(local_diag, t_diag);
+   }
+   else
+   {
+      t_diag.NewMemoryAndSize(local_diag.GetMemory(), length, false);
+   }
+   Solver *out = NULL;
+   if (chebyshev)
+   {
+      const int cheb_order = 3;
+      out = new OperatorChebyshevSmoother(&op, t_diag, ess_tdofs, cheb_order);
+   }
+   else
+   {
+      const double jacobi_scale = 0.65;
+      out = new OperatorJacobiSmoother(t_diag, ess_tdofs, jacobi_scale);
+   }
+   CeedVectorDestroy(&diagceed);
+   return out;
+}
+
+#ifdef MFEM_USE_MPI
+
+class CeedAMG : public Solver
+{
+public:
+   CeedAMG(ConstrainedMFEMCeedOperator &oper, HypreParMatrix *P)
+   {
+      MFEM_ASSERT(P != NULL, "");
+      const Array<int> ess_tdofs = oper.GetEssentialTrueDofs();
+      height = width = oper.Height();
+
+      CeedOperatorFullAssemble(oper.GetCeedOperator(), &mat_local);
+
+      {
+         HypreParMatrix hypre_local(
+            P->GetComm(), P->GetGlobalNumRows(), P->RowPart(), mat_local);
+         op_assembled = RAP(&hypre_local, P);
+      }
+      HypreParMatrix *mat_e = op_assembled->EliminateRowsCols(ess_tdofs);
+      delete mat_e;
+      amg = new HypreBoomerAMG(*op_assembled);
+      amg->SetPrintLevel(0);
+   }
+   void SetOperator(const Operator &op) { }
+   void Mult(const Vector &x, Vector &y) const { amg->Mult(x, y); }
+   ~CeedAMG()
+   {
+      delete op_assembled;
+      delete amg;
+      delete mat_local;
+   }
+private:
+   SparseMatrix *mat_local;
+   HypreParMatrix *op_assembled;
+   HypreBoomerAMG *amg;
+};
+
+#endif
+
+void CoarsenEssentialDofs(const Operator &interp,
+                          const Array<int> &ho_ess_tdofs,
+                          Array<int> &alg_lo_ess_tdofs)
+{
+   Vector ho_boundary_ones(interp.Height());
+   ho_boundary_ones = 0.0;
+   const int *ho_ess_tdofs_h = ho_ess_tdofs.HostRead();
+   for (int i=0; i<ho_ess_tdofs.Size(); ++i)
+   {
+      ho_boundary_ones[ho_ess_tdofs_h[i]] = 1.0;
+   }
+   Vector lo_boundary_ones(interp.Width());
+   interp.MultTranspose(ho_boundary_ones, lo_boundary_ones);
+   auto lobo = lo_boundary_ones.HostRead();
+   for (int i = 0; i < lo_boundary_ones.Size(); ++i)
+   {
+      if (lobo[i] > 0.9)
+      {
+         alg_lo_ess_tdofs.Append(i);
+      }
+   }
+}
+
+template <typename INTEG>
+int TryToAddCeedSubOperator(BilinearFormIntegrator *integ_in, CeedOperator op)
+{
+   INTEG *integ = dynamic_cast<INTEG*>(integ_in);
+   if (integ != NULL)
+   {
+      CeedCompositeOperatorAddSub(op, integ->GetCeedData()->oper);
+      return 1;
+   }
+   return 0;
```

Comment:
I think the easiest is to replace this function with two methods in `NonlinearFormIntegrator`: `SupportsCeed()` and `AddToCompositeOperator(...)` that throws an error by default, and to overload `SupportsCeed()` in the `Integrator`s that support libCEED:
```c++
virtual bool NonlinearFormIntegrator::SupportsCeed()
{
   return false;
}

bool MassIntegrator::SupportsCeed()
{
   return true; // Maybe DeviceCanUseCeed()
}

bool VectorMassIntegrator::SupportsCeed()
{
   return true;
}

bool ConvectionIntegrator::SupportsCeed()
{
   return true;
}

bool DiffusionIntegrator::SupportsCeed()
{
   return true;
}

bool VectorDiffusionIntegrator::SupportsCeed()
{
   return true;
}

bool VectorConvectionNLFIntegrator::SupportsCeed()
{
   return true;
}

void NonlinearFormIntegrator::AddToCompositeOperator(CeedOperator op) // We might want to use mfem::ceed::Operator
{
   if (SupportsCEED())
   {
      CeedCompositeOperatorAddSub(op, oper);
   }
   else
   {
      MFEM_ABORT("libCEED does not support this Integrator");
   }
}
```

Reply:
Actually, `AddToCompositeOperator(Integrator &integ, CeedOperator &op)` as a function and just adding `SupportsCeed()` to NonlinearFormIntegrator` might be enough.

Reply:
This is a very nice simplification, and I have implemented it. I am worried we just move the maintenance burden somewhere else (we have to remember to add `SupportsCeed()` to all the integrators when we add Ceed support) but it certainly looks much cleaner.